File: server/controllers/notificationController.js
import asyncHandler from 'express-async-handler';
import Notification from '../models/Notification.js';

// @desc   Get all notifications (paginated + filtered)
// @route  GET /api/notifications
export const getNotifications = asyncHandler(async (req, res) => {
  const { type, isRead, page = 1, limit = 20 } = req.query;

  const filter = {};
  if (req.user.role === 'admin') {
    filter.isAdminNotification = true;
  } else {
    filter.isAdminNotification = false;
    filter.userId = req.user._id;
  }
  if (type && type !== 'ALL') filter.type = type;
  if (isRead !== undefined) filter.isRead = isRead === 'true';

  const pageNum = Math.max(1, Number(page));
  const limitNum = Math.min(100, Number(limit));
  const skip = (pageNum - 1) * limitNum;

  const [notifications, total, unreadCount] = await Promise.all([
    Notification.find(filter)
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(limitNum)
      .populate('userId', 'name email'),
    Notification.countDocuments(filter),
    Notification.countDocuments({ ...filter, isRead: false }),
  ]);

  res.json({
    success: true,
    count: notifications.length,
    total,
    totalPages: Math.ceil(total / limitNum),
    currentPage: pageNum,
    unreadCount,
    data: notifications,
  });
});

// @desc   Mark all notifications as read
// @route  PUT /api/notifications/mark-read
export const markAllRead = asyncHandler(async (req, res) => {
  const filter =
    req.user.role === 'admin'
      ? { isRead: false, isAdminNotification: true }
      : { isRead: false, isAdminNotification: false, userId: req.user._id };
  await Notification.updateMany(filter, { isRead: true });
  res.json({ success: true, message: 'All notifications marked as read' });
});

// @desc   Mark single notification as read
// @route  PUT /api/notifications/:id/read
export const markOneRead = asyncHandler(async (req, res) => {
  const notification = await Notification.findById(req.params.id);
  if (!notification) {
    res.status(404);
    throw new Error('Notification not found');
  }

  if (req.user.role !== 'admin' && String(notification.userId) !== String(req.user._id)) {
    res.status(403);
    throw new Error('Not authorized to modify this notification');
  }

  notification.isRead = true;
  await notification.save();

  res.json({ success: true, data: notification });
});

// @desc   Get unread notification count
// @route  GET /api/notifications/unread-count
export const getUnreadCount = asyncHandler(async (req, res) => {
  const filter =
    req.user.role === 'admin'
      ? { isRead: false, isAdminNotification: true }
      : { isRead: false, isAdminNotification: false, userId: req.user._id };
  const count = await Notification.countDocuments(filter);
  res.json({ success: true, count });
});

File: backend/app/controllers/notification_controller.py
from app.core.exceptions import AppError
from app.models import collections
from app.services.notification_service import unread_count_for_actor
from app.utils.bson_utils import serialize_doc, serialize_many, to_object_id
from app.utils.response import paginated_response
from app.utils.datetime_utils import utc_now


async def get_notifications_controller(db, current_user, query):
    filter_query = {}
    user_oid = to_object_id(current_user["id"])

    if current_user["role"] == "admin":
        filter_query["isAdminNotification"] = True
    else:
        filter_query["isAdminNotification"] = False
        filter_query["userId"] = user_oid

    if query.type and query.type != "ALL":
        filter_query["type"] = query.type
    if query.isRead is not None:
        filter_query["isRead"] = query.isRead

    page = query.page
    limit = query.limit
    skip = (page - 1) * limit

    cursor = db[collections.NOTIFICATIONS].find(filter_query).sort("createdAt", -1).skip(skip).limit(limit)
    notifications = await cursor.to_list(length=limit)
    total = await db[collections.NOTIFICATIONS].count_documents(filter_query)
    unread = await db[collections.NOTIFICATIONS].count_documents({**filter_query, "isRead": False})

    return paginated_response(
        data=serialize_many(notifications),
        total=total,
        page=page,
        limit=limit,
        extra={"unreadCount": unread},
    )


async def mark_all_read_controller(db, current_user):
    query = {"isRead": False}
    if current_user["role"] == "admin":
        query["isAdminNotification"] = True
    else:
        query["isAdminNotification"] = False
        query["userId"] = to_object_id(current_user["id"])

    await db[collections.NOTIFICATIONS].update_many(query, {"$set": {"isRead": True, "updatedAt": utc_now()}})
    return {"success": True, "message": "All notifications marked as read"}


async def mark_one_read_controller(db, notification_id: str, current_user):
    notification = await db[collections.NOTIFICATIONS].find_one({"_id": to_object_id(notification_id)})
    if not notification:
        raise AppError(status_code=404, message="Notification not found")

    if current_user["role"] != "admin" and str(notification.get("userId")) != current_user["id"]:
        raise AppError(status_code=403, message="Not authorized to modify this notification")

    await db[collections.NOTIFICATIONS].update_one(
        {"_id": notification["_id"]},
        {"$set": {"isRead": True, "updatedAt": utc_now()}},
    )
    updated = await db[collections.NOTIFICATIONS].find_one({"_id": notification["_id"]})
    return {"success": True, "data": serialize_doc(updated)}


async def get_unread_count_controller(db, current_user):
    count = await unread_count_for_actor(
        db,
        role=current_user["role"],
        user_object_id=to_object_id(current_user["id"]),
    )
    return {"success": True, "count": count}